docs/design/ide_kpm_watcher.md

```diff
@@ -0,0 +1,78 @@
+## KCL Watch System
```

Comment:
Can you list out the core interfaces of KCL Watch System? How external systems use it?

Reply:
Other watch systems interact via CLI ( using --watch flag in tsc, pyright, etc ) where it detects changes to a file and compile and run automatically.

Reply:
I am not sure about the VS code extension integration of this watch as I couldnt find any strong documentation about that for other systems( maybe I am missing something if it possible can you help find the integration document of it? ).

But IMO we should start with KCL --watch CLI interface first.

Reply:
Note that other systems here also include the kcl language server. You can directly describe these integrations in the design document.

Reply:
The thing which im currently stuck at is "how other systems integrated their watch system in their respective LSP" because when I went to their official documentation apart from getting relevant info about the CLI implementation use case, I couldn't find how it's integrated in LSP.

Reply:
Can you help here a bit, if possible?

Reply:
In this design, simply describe clearly how to integrate the watch into the kcl language server. LSP can now return changes to the file, and we need to integrate the corresponding logic to complete the corresponding functions.

Reply:
We can do that in the following ways as far as I have read the code of the KCL in its LSP folder,

When initializing the KCL language server, we can create a new instance of the Watcher component. This can be done at the startup of the language server.
Something like this:
```rust
let watcher = Watcher::new();
```
After that we can Load configuration settings from a configuration file (e.g., `kcl.toml`) and Define watch paths and supported file types in the configuration.
we can take help of this code https://github.com/RLado/cfg_mgr that can help us load that file.

Then we can register watch paths with the Watcher to monitor file system events and use the LSP to obtain the workspace folders and register them as watch paths.

When a file system event occurs within a watched path, the Watcher detects the event and uses LSP's `didChange` notification to capture file changes and handle them accordingly.

```rust
fn handle_file_change(uri: String) {
    // Detect file type based on extension
    let file_type = FileTypeDetector::detect(uri);

    // Invoke corresponding handler
    match file_type {
        FileType::K => KHandler::handle(uri),
        FileType::Mod => ModHandler::handle(uri),
        FileType::JSON => JSONHandler::handle(uri),
        FileType::YAML => YAMLHandler::handle(uri),
    }
}
```

We can utilize LSP's `DidChangeWatchedFiles` notification to receive file change events from the client and update the `handle_file_change` function to integrate with LSP's file change notifications.

Reply:
I am studying the LSP folder code, also wanted to ask do I need to look up to any other folder?

Reply:
Just the LSP folder